File: greedy/763-partitionLabels.cpp
#include<iostream>
#include<vector>
#include<string>
#include<cmath>

using namespace std;

using VEC=std::vector<int>;

class Solution {
public:
    VEC partitionLabels(const string& s) {
        int n=s.size();
        VEC ans{};
        VEC lastIndex(128,0);
        for(int i=0;i<n;++i)
        {
            lastIndex[s[i]]=i;
        }

        int start=0,end=0;
        for(int i=0;i<n;++i)
        {
            end=max(end,lastIndex[s[i]]);
            if(end==i)
            {
                ans.push_back(end-start+1);
                start=end+1;  //here should update the start as end+1
            }
        }
        return ans;
    }
};

int main()
{
    string s{"ababcbacadefegdehijhklij"};
    vector<int> res=Solution().partitionLabels(s);
    for(auto v:res)
        cout<<v<<',';
}

File: greedy/763-partitionLabels.py


def partitionLabels(s):
    n=len(s)
    lastIndex=[0]*128
    ans=[]
    start,end=0,0
    for i in range(0,n):
        lastIndex[ord(s[i])]=i
    
    for i in range(0,n):
        end=max(end,lastIndex[ord(s[i])])
        if end==i:
            ans.append(end-start+1)
            start=end+1
    return ans

s="ababcbacadefegdehijhklij"
print(partitionLabels(s))